server/ai_live_video.go

```diff
@@ -1,33 +1,52 @@
 package server
 
 import (
+	"context"
 	"errors"
 	"fmt"
 	"io"
 	"log/slog"
 	"net/url"
 	"os"
+	"time"
 
 	"github.com/livepeer/go-livepeer/core"
 	"github.com/livepeer/go-livepeer/media"
 	"github.com/livepeer/go-livepeer/trickle"
 	"github.com/livepeer/lpms/ffmpeg"
 )
 
-func startTricklePublish(url *url.URL, params aiRequestParams) {
+func startTricklePublish(url *url.URL, params aiRequestParams, sess *AISession, mid string) {
 	publisher, err := trickle.NewTricklePublisher(url.String())
 	if err != nil {
 		slog.Info("error publishing trickle", "err", err)
 	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	paymentProcessInterval := 1 * time.Second
+	paymentSender := livePaymentSender{segmentsToPayUpfront: 5}
+	f := func(inPixels int64) error {
+		return paymentSender.SendPayment(context.Background(), &SegmentInfoSender{
+			sess:      sess.BroadcastSession,
+			inPixels:  inPixels,
+			priceInfo: sess.OrchestratorInfo.PriceInfo,
+			mid:       mid,
+		})
+	}
+	paymentProcessor := NewLivePaymentProcessor(ctx, paymentProcessInterval, f)
+
 	params.segmentReader.SwitchReader(func(reader io.Reader) {
 		// check for end of stream
 		if _, eos := reader.(*media.EOSReader); eos {
 			if err := publisher.Close(); err != nil {
 				slog.Info("Error closing trickle publisher", "err", err)
 			}
+			cancel()
 			return
 		}
 		go func() {
+			reader := paymentProcessor.process(reader)
```

Comment:
@j0sh Is it how you imagined tweaking into the segment processing logic?

Reply:
That seems fine to me, yes. 1s cadence is probably too frequent though - segments are usually longer than that, eg 2-3s keyframe interval.

Now that I am connecting the bits here mentally while reviewing this ... returning a reader via the API is an interesting choice because it does force us to tee the input reader and handle the returned reader on the API side even if there is nothing else to do with it, eg on the orchestrator where we only discard it.

I don't really have a better idea right now for how to do this, other than making the "should skip" function as part of the API and calling `process` only periodically, doing tee outside the API as necessary. Kind of boilerplate-y though.